src/core/maprenderer/qgsmaprendererjob.cpp

```diff
@@ -725,47 +812,75 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
-      continue;
+      job2.context()->setPainter( allocateImageAndPainter( job.layerId, job2.img ) );
+    }
+    else
+    {
+      PictureAndPainter pictureAndPainter = allocatePictureAndPainter();
+      job2.imgPic.reset( pictureAndPainter.first );
+      job2.context()->setPainter( pictureAndPainter.second );
     }
 
-    // Points to the first pass job. This will be needed during the second pass composition.
-    for ( MaskSource &source : sourceList )
+    if ( ! job2.img && ! job2.imgPic )
     {
-      if ( source.labelMaskId != -1 )
-        job2.maskJobs.push_back( qMakePair( nullptr, source.labelMaskId ) );
-      else
-        job2.maskJobs.push_back( qMakePair( layerJobMapping[source.layerId], -1 ) );
+      secondPassJobs.pop_back();
+      continue;
     }
 
     // FIXME: another possibility here, to avoid allocating a new map renderer and reuse the one from
     // the first pass job, would be to be able to call QgsMapLayerRenderer::render() with a QgsRenderContext.
-    QgsVectorLayerRenderer *mapRenderer = static_cast<QgsVectorLayerRenderer *>( vl1->createMapRenderer( *job2.context() ) );
+    QgsVectorLayerRenderer *mapRenderer = static_cast<QgsVectorLayerRenderer *>( ml->createMapRenderer( *job2.context() ) );
     job2.renderer = mapRenderer;
     if ( job2.renderer )
     {
       job2.context()->setFeedback( job2.renderer->feedback() );
     }
 
-    // Modify the render context so that symbol layers get disabled as needed.
-    // The map renderer stores a reference to the context, so we can modify it even after the map renderer creation (what we need here)
-    job2.context()->setDisabledSymbolLayers( QgsSymbolLayerUtils::toSymbolLayerPointers( mapRenderer->featureRenderer(), symbolList ) );
+    if ( !forceVector || job2.needRasterization )
+    {
+      // Render only the non masked symbol layer and we will compose 2nd pass with mask and first pass
+      // rendering in composeSecondPass
+      // The map renderer stores a reference to the context, so we can modify it even after the map renderer creation (what we need here)
+      job2.context()->setDisabledSymbolLayers( QgsSymbolLayerUtils::toSymbolLayerPointers( mapRenderer->featureRenderer(), symbolList ) );
+    }
+    else
+    {
+      const QSet<const QgsSymbolLayer *> &symbolLayers = QgsSymbolLayerUtils::toSymbolLayerPointers( mapRenderer->featureRenderer(), symbolList );
+
+      // we set clip path on masked symbol layer and resulting 2nd pass job picture will be the final
+      // rendering
+      for ( QPair<LayerRenderJob *, int> p : job2.maskJobs )
+      {
+        QPainter *maskPainter = p.first ? p.first->context()->maskPainter() : labelJob.context.maskPainter();
+        const QPainterPath &path = static_cast<QgsMaskPaintDevice *>( maskPainter->device() )->maskPainterPath();
+        for ( const QgsSymbolLayer *symbolLayer : symbolLayers )
+        {
+          job2.context()->addSymbolLayerClipPath( symbolLayer, &path );
```

Comment:
I'm not super-convinced by setting this upfront here. It would be much cleaner to move this logic to happen after the first pass job is completed (and the painter paths have been completely calculated), just before we start the second pass.

Reply:
Fixed [a263d82](https://github.com/qgis/QGIS/pull/47459/commits/a263d82797116bea7d0517b44dd01c2e414d7c33)